packages/frontend/src/redux/reducers/sign/index.js

```diff
@@ -25,43 +30,62 @@ const sign = handleActions({
                 .reduce((totalAmount, amount) => totalAmount.add(new BN(amount)), new BN(0)).toString(),
             fees: {
                 transactionFees: '', // TODO: Calculate total fees
-                gasLimit: allActions
-                    .filter(a => Object.keys(a)[0] === 'functionCall')
-                    .map(a => a.functionCall.gas)
-                    .reduce((totalGas, gas) => totalGas.add(gas), new BN(0)).toString(),
+                gasLimit: calculateGasLimit(allActions),
                 gasPrice: '' // TODO: Where to get gas price?
             },
             sensitiveActionsCounter: allActions
                 .filter(a => ['deployContract', 'stake', 'deleteAccount'].indexOf(Object.keys(a)[0]) > -1)
                 .length
         };
     },
-    [signAndSendTransactions]: (state, { error, payload, ready }) => {
-
-        if (!ready) {
-            return {
-                ...state,
-                status: 'needs-confirmation'
-            };
-        }
+    [handleSignTransactions.pending]: (state) => {
+        const { retryTxDirection, status } = state;
 
-        if (error) {
-            return {
-                ...state,
-                status: 'error',
-                error: payload
-            };
-        }
+        const transactions = status === SIGN_STATUS.RETRY_TRANSACTION
+            ? increaseGasForTransactions({ 
+                transactions: cloneDeep(state.transactions),
+                retryTxDirection: state.retryTxDirection || retryTxDirection }
+            )
+            : state.transactions;
 
         return {
             ...state,
-            status: 'success'
+            status: SIGN_STATUS.IN_PROGRESS,
+            transactions,
+            fees: {
+                ...state.transactions.fees,
+                gasLimit: calculateGasLimit(transactions.flatMap(t => t.actions))
+            }
         };
     },
-    [setSignTransactionStatus]: (state, { payload }) => {
+    [handleSignTransactions.fulfilled]: (state, { payload }) => ({
+        ...state,
+        status: SIGN_STATUS.SUCCESS,
+        transactionHashes: payload,
+        error: undefined
+    }),
+    [handleSignTransactions.rejected]: (state, { error }) => {
+        const retryTxDirection = error.message.includes('Exceeded the prepaid gas')
+            ? RETRY_TX.INCREASE
+            : undefined;
+        
+        const tryRetryTx = retryTxDirection && !state.transactions.every((t) => 
+            t.actions && t.actions.every((a) => a.functionCall && a.functionCall.gas && (
+                state.retryTxDirection === RETRY_TX.INCREASE 
+                && (
+                    a.functionCall.gas.gt(new BN(RETRY_TX.GAS.MAX))
+                    || a.functionCall.gas.eq(new BN(RETRY_TX.GAS.MAX))
+                )
+            ))
+        );
```

Comment:
1. Please don't create so many levels of nested expressions; they are hard to read and difficult to reason about; they tend to hide redundancy and make code non-linear -- please expand your expressions into function leveraging early excit/short-circuits and consider readability as a priority (add comments describing the WHY for code we are writing)
2. Please remove the unused complexity of using `retryTxDirection` here.
3. `BN.js` has a `.gte` method; no need to use an OR here, but the check can also be simplified by removing the negation and using `BN.lt()` instead.
4. This current code will treat actions that are _not_ function call type actions as a reason to not allow retrying of the parent transaction, even though there may be only non-function call actions and some function call actions that can have their gas increased; non-functionCall actions shouldn't be considered when computing `tryRetryTx`
5. Per my prior comments, _what will happen if some of the transactions are successfully submitted and the user clicks 'Retry' at the top level_?  Will we be double-submitting the same transactions for the user when they click 'retry'?  If so, we are missing some state management; we should either mark each transaction as we successfully send them so that we can filter them out from the retry logic, or disable the user's ability to click Retry if any of the transactions have been successfully submitted.
6. Rather than negating `every()`, please use `.some()` and use affirmitive logic -- it's more efficient (it exits as soon as one element matches the condition rather than iterating all the way to the end), and is idiomatic language (immediately recognizable grammar).

Something addressing # 1- # 4 above might look something like:
```javascript
const notEnoughGasAttached = error.message.includes('Exceeded the prepaid gas');
const hasAtLeastOneFunctionCallAction = state.transactions.some((t) => {
    return (t.actions || []).some((a) => a && a.functionCall);
});

const canRetryWithIncreasedGas = notEnoughGasAttached && hasAtLeastOneFunctionCallAction && state.transactions.some((t) => {
        if (!t.actions) { return false; }

        return t.actions.some((a) => {
            // We can only increase gas for actions that are function calls and still have < RETRY_TX.GAS.MAX gas allocated
            if (!a || !a.functionCall) { return false; }

            return a.functionCall.gas.lt(new BN(RETRY_TX.GAS.MAX));
        });
    }
);
```

Reply:
1. You're right, force of habit :) Will keep that in mind.
2. Done.
3. Done.
4. Done.
5. Will post the answer in the previous comment, that I missed.
6. Done.
Thanks for the example.